lib/mongo/collection/view/change_stream.rb

```diff
@@ -0,0 +1,143 @@
+        #
+        # @since 2.5.0
+        attr_reader :options
+
+        # Initialize the change stream for the provided collection view, pipeline
+        # and options.
+        #
+        # @example Create the new change stream view.
+        #   ChangeStream.new(view, pipeline, options)
+        #
+        # @param [ Collection::View ] view The collection view.
+        # @param [ Array<Hash> ] pipeline The pipeline of operators to filter the change notifications.
+        # @param [ Hash ] opts The change stream options.
+        #
+        # @option options [ String ] :full_document Allowed values: ‘default’, ‘updateLookup’. Defaults to ‘default’.
+        #   When set to ‘updateLookup’, the change notification for partial updates will include both a delta
+        #   describing the changes to the document, as well as a copy of the entire document that was changed
+        #   from some time after the change occurred.
+        # @option options [ BSON::Document, Hash ] :resume_after Specifies the logical starting point for the
+        #   new change stream.
+        # @option options [ Integer ] :max_await_time_ms The maximum amount of time for the server to wait
+        #   on new documents to satisfy a change stream query.
+        # @option options [ Integer ] :batch_size The number of documents to return per batch.
+        # @option options [ BSON::Document, Hash ] :collation The collation to use.
+        #
+        # @since 2.5.0
+        def initialize(view, pipeline, options = {})
+          @view = view
+          @change_stream_filters = pipeline.dup
+          @options = options.dup.freeze
+          @resume_token = @options[:resume_after]
+          read_with_one_retry { create_cursor! }
+        end
+
+        # Iterate through documents returned by the change stream.
+        #
+        # @example Iterate through the stream of documents.
+        #   stream.each do |document|
+        #     p document
+        #   end
+        #
+        # @return [ Enumerator ] The enumerator.
+        #
+        # @since 2.5.0
+        #
+        # @yieldparam [ BSON::Document ] Each change stream document.
+        def each
+          @cursor.each do |doc|
+            cache_resume_token(doc)
+            yield doc
+          end if block_given?
+          @cursor.to_enum
+        rescue => e
+          close
+          if retryable?(e)
+            create_cursor!
+            retry
+          end
+          raise
+        end
+
+        # Close the change stream.
+        #
+        # @example Close the change stream.
+        #   stream.close
+        #
+        # @return [ nil ] nil.
+        #
+        # @since 2.5.0
+        def close
+          if @cursor
+            begin; @cursor.send(:kill_cursors); rescue; end
+            @cursor = nil
+          end
+        end
+
+        def closed?
+          !!@cursor
+        end
+
+        private
+
+        def cache_resume_token(doc)
+          unless @resume_token = doc[:_id]
```

Comment:
Similarly, do you need to duplicate the _id to prevent a user from modifying it?

Reply:
No, because ObjectIds are immutable and we are caching the ObjectId itself, not the document. If we were caching the document, then we might have to duplicate it. Thanks for checking though.

Reply:
The `_id` here is a change stream resume token which is a document.

Reply:
Oh right, I forgot. hm, I guess I should duplicate it then. Thanks!